### PR TITLE
raftstore: Implement observer on_compute_engine_size

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What is changed and how it works?

Issue Number: ref #12849

What's Changed:

```commit-message
Implement observer on_compute_engine_size
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/coprocessor/mod.rs [hunk 1]
```diff
@@ -151,6 +151,18 @@ pub trait SplitCheckObserver<E>: Coprocessor {
     );
 }
 
+#[derive(Debug, Default)]
+pub struct EngineSize {
```
Comment (tonyxuqqi) [inline components/raftstore/src/coprocessor/mod.rs:155]: Is it better to rename EngineSize as StoreSizeInfo?

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore/src/coprocessor/mod.rs [hunk 2]
```diff
@@ -203,6 +203,18 @@ pub trait SplitCheckObserver<E>: Coprocessor {
     );
 }
 
+#[derive(Debug, Default)]
+pub struct StoreSizeInfo {
+    pub capacity: u64,
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/mod.rs:208]: Please add some comments.

File: components/raftstore/src/store/worker/pd.rs [hunk 3]
```diff
@@ -2304,6 +2284,49 @@ fn collect_report_read_peer_stats(
     stats
 }
 
+fn collect_engine_size<EK: KvEngine, ER: RaftEngine>(
+    coprocessor_host: &CoprocessorHost<EK>,
+    store_info: Option<&StoreInfo<EK, ER>>,
+    snap_mgr_size: u64,
+) -> Option<(u64, u64, u64)> {
+    if let Some(engine_size) = coprocessor_host.on_compute_engine_size() {
+        Some((engine_size.capacity, engine_size.used, engine_size.avail))
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/pd.rs:2293]: How about returning? So there will be no indentation in other branch.

File: components/raftstore/src/coprocessor/mod.rs (hunk 2 above)
Comment (BusyJay) [reply components/raftstore/src/coprocessor/mod.rs:209]: And what's the meaning of all these fields? Why there is a `used` when the other two are available?

**Review by BusyJay — APPROVED** [review]

**Pull request merged**